### ISSUE TITLE
Optimize `avoid_unused_parameters` for Code Generators

### ISSUE THREAD

**Issue by nathnaeld** [charlescyt/pyramid_lint#69]
### Feature Type

Lint

### Feature Description

The `avoid_unused_parameters` lint rule often flags parameters in classes generated by packages like `freezed` as "unused." These parameters are crucial for the generated code but are not directly used within the class. This leads to false positives and forces to manually suppress the lint, adding unnecessary code noise.

## Proposed Solution

Update the `avoid_unused_parameters` lint rule to allow ignoring parameters in constructors of classes that have specific annotations. This can be configured in the `custom_lint` section, allowing developers to specify which annotations should bypass the check.

**Eample:**
```yaml
custom_lint:
  rules:
    - avoid_unused_parameters:
      ignored_parameters: ['ref']
      ignored_annotations: ['freezed', 'riverpod']
```

## Current Behavior
```dart
@freezed
abstract class User with _$User {
  const factory User({
    // Linter flags 'name' and 'age' as unused
    required String name,
    required int age,
  }) = _User;
}
```

**Comment by charlescyt** [comment]
Thanks for the suggestion! I will work on it this week.

**Comment by nathnaeld** [comment]
> Could you check again which lint was triggered in the `freezed` snippet? `avoid_unused_parameters` only checks parameters of `FunctionDeclaration`s and `MethodDeclaration`s. Therefore it should not lint any unused parameters of a `ConstructorDeclaration`s.

Thanks for looking into this. I upgraded to Flutter 3.35 yesterday, and I’ve noticed that after SDK upgrades, lint indications aren’t always consistent. Right now, I can’t reproduce the issue anymore — the lint warning is gone. Previously, I was suppressing it with `avoid_unused_parameters` on mostly named constructors. Could we keep this open for a little while, just in case it resurfaces?

**Comment by charlescyt** [comment]
Thanks for the clarification!

Regarding your suggestion, there is already an official lint available: [`avoid_unused_constructor_parameters`](https://dart.dev/tools/linter-rules/avoid_unused_constructor_parameters). I would rather not implement a lint that overlaps with the official one. If the official lint does not cover your use case, feel free to let me know, and I would be happy to have a further discussion.

Though I agree the name `avoid_unused_parameters` is a bit misleading, a better name could be `avoid_unused_function_parameters`.
